### PR TITLE
Add structref as a mutable struct that is pass-by-ref

### PR DESCRIPTION
**Pull request by sklam**
Closes https://github.com/numba/numba/issues/5397

Add `structref` to support mutable structure that can be extended like typical custom types (i.e. with `@overload` and friends). Also, support some nice default to minimizing work; i.e. default boxer/unboxer, default getter/setter, default constructor.

TODOs:

- add docs. to be included in a different PR.

### CODE REVIEW

File: numba/tests/test_struct_ref.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+
+# Call to define_proxy is needed to register the use of `MyStruct` as a
+# PyObject proxy for creating a Numba-allocated structref.
+# The `MyStruct` class and then be used in both jit-code and interpreted-code.
+structref.define_proxy(
+    MyStruct,
+    MyStructType,
+    ['values', 'counter'],
+)
```
Comment (stuartarchibald) [inline numba/tests/test_struct_ref.py:72]: Is there benefit in making this into a factory? Also wonder if the getters and setters like those here https://github.com/numba/numba/pull/5978/files#diff-9b2fbc1beb945aef47f6c71220bcd54dR101-R113 could be autogenerated in such a case, perhaps into a base class from which the other's can inherit and later be proxied? Perhaps not for now :)!

File: numba/tests/test_struct_ref.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        self.assertEqual(td['a'].values, 1)
+        self.assertEqual(td['a'].counter, 2.3)
+        # overwrite
+        td['a'] = MyStruct(2, 3.3)
+        self.assertEqual(td['a'].values, 2)
+        self.assertEqual(td['a'].counter, 3.3)
+        # mutate
+        td['a'].values += 10
+        self.assertEqual(td['a'].values, 12)    # changed
+        self.assertEqual(td['a'].counter, 3.3)  # unchanged
```
Comment (stuartarchibald) [inline numba/tests/test_struct_ref.py:185]: Perhaps also check a `td['b'] = MyStruct(4, 5.6)` works, as I think it should? Makes me wonder, should a `MyStruct(int8(4), 5.6)` be castable, I'd guess not at the typing layer?

File: numba/tests/test_struct_ref.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+@overload_method(MyStructType, "testme")
+def _(self, arg):
+    def impl(self, arg):
+        return self.values * arg + self.counter
+    return impl
+
+
+class TestStructRefExtending(MemoryLeakMixin, TestCase):
+    def test_overload_method(self):
```
Comment (stuartarchibald) [inline numba/tests/test_struct_ref.py:211]: RE overloads, is attribute overloading likely to ever be a thing that makes sense in the `structref` context?

File: docs/source/developer/repomap.rst [hunk 4]
```diff
@@ -82,6 +82,7 @@ These define aspects of the public Numba interface.
   regular functions on the CPU
 - :ghfile:`numba/core/extending.py` - Public decorators for extending Numba
   (``overload``, ``intrinsic``, etc)
+  - :ghfile:`numba/core/structref.py` - Public API for defining mutable struct
```
Comment (stuartarchibald) [inline docs/source/developer/repomap.rst:85]:
```suggestion
  - :ghfile:`numba/core/structref.py` - Public API for defining a mutable struct
```

File: numba/core/structref.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        field_type = inst_type.field_dict[attr]
+        casted = context.cast(builder, val, val_type, field_type)
+        # read old
+        old_value = getattr(dataval, attr)
+        # incref new value
+        context.nrt.incref(builder, val_type, casted)
+        # decref old value (must be last in case new value is old value)
+        context.nrt.decref(builder, val_type, old_value)
+        # write new
+        setattr(dataval, attr, casted)
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/structref.py:125]: I thought that would happen too. But, it turns out that the typer for `setattr` is forcing the RHS to match the attribute type.

File: numba/core/structref.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
+"""Utilities for defining a mutable struct.
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/structref.py:1]: Good point. Better be in experimental!

File: numba/tests/test_struct_ref.py (hunk 1 above)
Comment (sklam) [reply numba/tests/test_struct_ref.py]: I was trying to autogenerate the getters/setters via a base class. Just too much work for now.

File: numba/tests/test_struct_ref.py (hunk 3 above)
Comment (sklam) [reply numba/tests/test_struct_ref.py:211]: I guess to emulate `@property` without an actual field backing it. I can add a test quickly

File: numba/tests/test_struct_ref.py (hunk 2 above)
Comment (sklam) [reply numba/tests/test_struct_ref.py:185]: It's not casting the fields at the moment. That would be another potential enhancement

File: docs/source/developer/repomap.rst [hunk 7]
```diff
@@ -82,7 +82,7 @@ These define aspects of the public Numba interface.
   regular functions on the CPU
 - :ghfile:`numba/core/extending.py` - Public decorators for extending Numba
   (``overload``, ``intrinsic``, etc)
-  - :ghfile:`numba/core/structref.py` - Public API for defining mutable struct
+  - :ghfile:`numba/core/structref.py` - Public API for defining a mutable struct
```
Comment (stuartarchibald) [inline docs/source/developer/repomap.rst:85]:
```suggestion
  - :ghfile:`numba/experimental/structref.py` - Public API for defining a mutable struct
```
should fix the doc build.

File: numba/core/types/containers.py [hunk 8]
```diff
@@ -722,18 +722,28 @@ def __init__(self, fields):
...
             name, and `type` is a numba type for the field type.
         """
-        self._fields = tuple(map(tuple, fields))
+        def check_field_pair(fieldpair):
+            name, typ = fieldpair
+            if not isinstance(name, str):
+                msg = "expecting a str for field name"
+                raise ValueError(msg)
+            if not isinstance(typ, Type):
+                msg = f"expecting a Numba Type for field type"
```
Comment (stuartarchibald) [inline numba/core/types/containers.py:735]:
```suggestion
                msg = "expecting a Numba Type for field type"
```

**Comment by stuartarchibald** [comment]
Thanks for the fixes, I think the only remaining this is the `f-string` problem here https://github.com/numba/numba/pull/5978#discussion_r455201605 else this looks good to go.

**Pull request merged**